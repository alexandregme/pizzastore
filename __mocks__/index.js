export const TOPPINGS = [{
  name:"topping 1",
  price:1,
  isChecked:true
},{
  name:"topping 2",
  price:2,
  isChecked:false
},{
  name:"topping 3",
  price:3,
  isChecked:false
},{
  name:"topping 4",
  price:4,
  isChecked:false
},{
  name:"topping 5",
  price:5,
  isChecked:false
},{
  name:"topping 6",
  price:6,
  isChecked:false
},{
  name:"topping 7",
  price:7,
  isChecked:false
},{
  name:"topping 8",
  price:8,
  isChecked:false
}];

export const DEFAULT_TOPPING = {
  name:"default topping",
  price:0,
  isChecked:false
};

export const CUSTOM_TOPPING = {
  name:"custom topping",
  price:1,
  isChecked:true
};

export const DEFAULT_PIZZA = {
  "name": "default pizza",
  "maxToppings": 0,
  "basePrice": 0,
  "totalPrice": 0,
  toppings:[]
};

export const CUSTOM_PIZZA = {
  "name": "custom pizza",
  "maxToppings": 1,
  "basePrice": 1,
  "totalPrice": 1,
  toppings:[{
    name:"topping 1",
    price:1,
    isChecked:true
  }]
};

export const PIZZAS = [{
  "name": "small",
  "maxToppings": 3,
  "basePrice": 9.89,
  toppings: TOPPINGS
},{
  "name": "medium",
  "maxToppings": 5,
  "basePrice": 10.89,
  toppings: TOPPINGS
}, {
  "name": "large",
  "maxToppings": null,
  "basePrice": 13.49,
  toppings: TOPPINGS
}];
