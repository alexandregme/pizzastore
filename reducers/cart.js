let nextid = 0;


const addItemToCart = (state = [], action) => {
  return [
    ...state,
    {
      id: nextid++,
      ...action.pizza
    }
  ];
};

const removeItemFromCart = (state = [], action) => {
  return state.filter((item)=> item.id !== action.id);
};

const totalCostCart = (state = []) => {
  return  state.reduce((sum, item) => sum + item.totalPrice, 0);
};

const cart = (state = [], action) => {
  let modifiedCart = {};

  switch (action.type) {
    case 'ADD_TO_CART':
      modifiedCart.items = addItemToCart(state.items, action);
      modifiedCart.totalCost = totalCostCart(modifiedCart.items);
      modifiedCart.totalItems = modifiedCart.items.length;
      return modifiedCart;

    case 'REMOVE_FROM_CART':
      modifiedCart.items = removeItemFromCart(state.items, action);
      modifiedCart.totalCost = totalCostCart(modifiedCart.items);
      modifiedCart.totalItems = modifiedCart.items.length;
      return modifiedCart;

    default:
      return state;
  }
};

export default cart;
