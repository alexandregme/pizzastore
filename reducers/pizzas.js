const pizzas = (state = [], action) => {
  switch (action.type) {
    case 'FETCH_PIZZAS':
      return action.pizzas;
    case 'TOGGLE_TOPPING_FROM_PIZZA':
      return state.map(pizza =>{
        if (pizza.name !== action.pizzaName) {
          return pizza;
        }else {
          return {
            ...pizza,
            toppings: pizza.toppings.map(topping => {
              if (topping.name !== action.toppingName) {
                return topping
              }
              return {
                ...topping,
                isChecked: !topping.isChecked
              }
            })
          }
        }
      });
    default:
      return state;
  }
};

export default pizzas;
