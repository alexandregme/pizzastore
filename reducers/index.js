import { combineReducers } from 'redux';
import pizzas from './pizzas';
import cart from './cart';

export default combineReducers({pizzas, cart});
