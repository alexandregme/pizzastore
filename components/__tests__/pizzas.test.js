import React from 'react';
import { shallow } from 'enzyme';
import { createStore } from "redux";
import reducers from "../../reducers";
import expect from 'expect';
import ConnectedPizzas, { Pizzas } from '../pizzas';
import Pizza from '../pizza';
import {PIZZAS, STORE} from "../../__mocks__";

describe("Pizzas", ()=>{
  let mountedPizzas;
  beforeEach(()=>{
    mountedPizzas = shallow(<Pizzas />);
  });

  it('renders without crashing', () => {
    shallow(<Pizzas />);
    expect(mountedPizzas).toMatchSnapshot();
  });

  it('expect to show no pizzas found msg', () => {
    let notFoundMsg = mountedPizzas.find('h2');
    expect(notFoundMsg).toHaveLength(1);
    expect(notFoundMsg.text()).toEqual('No pizzas were found, plz come back soon.');
  });
});

describe("Pizzas Custom component props", ()=>{
  let mountedPizzas;
  let props = { pizzas: PIZZAS };

  beforeEach(()=>{
    mountedPizzas = shallow(<Pizzas {...props} />);
  });

  it('expect to render 3 pizza sizes', () => {
    let pizzas = mountedPizzas.find(Pizza);
    expect(pizzas.length).toEqual(3);
  });
});

describe('Connected Pizzas', () => {
  let connectedPizzas;
  let wrapper;

  beforeEach(() => {
    const store = createStore(reducers, STORE);
    wrapper = shallow(<ConnectedPizzas store={store} />);
    connectedPizzas = wrapper.find(Pizzas);
  });

  it('should have pizzas equals to mock consumed from store', () => {
    expect(connectedPizzas.props().pizzas).toBe(PIZZAS);
  });
});
