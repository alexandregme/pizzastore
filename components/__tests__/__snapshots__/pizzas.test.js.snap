// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pizzas renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pizzas />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="card-deck"
      >
        <h2>
          No pizzas were found, plz come back soon.
        </h2>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h2>
          No pizzas were found, plz come back soon.
        </h2>,
        "className": "card-deck",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No pizzas were found, plz come back soon.",
        },
        "ref": null,
        "rendered": "No pizzas were found, plz come back soon.",
        "type": "h2",
      },
      "type": "div",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="card-deck"
        >
          <h2>
            No pizzas were found, plz come back soon.
          </h2>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h2>
            No pizzas were found, plz come back soon.
          </h2>,
          "className": "card-deck",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No pizzas were found, plz come back soon.",
          },
          "ref": null,
          "rendered": "No pizzas were found, plz come back soon.",
          "type": "h2",
        },
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
